proposals/p2006.md

```diff
@@ -0,0 +1,654 @@
+    -   [Indexing](#indexing)
+    -   [`const`-qualified types](#const-qualified-types)
+    -   [Interop with `const &`](#interop-with-const-)
+    -   [Customization](#customization)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Immutable value escape hatch](#immutable-value-escape-hatch)
+    -   [References in addition to pointers](#references-in-addition-to-pointers)
+    -   [Automatic dereferencing](#automatic-dereferencing)
+    -   [Exclusively using references](#exclusively-using-references)
+    -   [Alternative syntaxes for locals](#alternative-syntaxes-for-locals)
+-   [Appendix: background, context, and use cases from C++](#appendix-background-context-and-use-cases-from-c)
+    -   [`const` references versus `const` itself](#const-references-versus-const-itself)
+    -   [Pointers](#pointers)
+    -   [References](#references)
+        -   [Special but critical case of `const T&`](#special-but-critical-case-of-const-t)
+        -   [R-value references and forwarding references](#r-value-references-and-forwarding-references)
+        -   [Mutable operands to user-defined operators](#mutable-operands-to-user-defined-operators)
+        -   [User-defined dereference and indexed access syntax](#user-defined-dereference-and-indexed-access-syntax)
+        -   [Member and subobject accessors](#member-and-subobject-accessors)
+        -   [Non-null pointers](#non-null-pointers)
+        -   [Syntax-free dereference](#syntax-free-dereference)
+
+<!-- tocstop -->
+
+## Abstract
+
+Introduce a concrete design for how Carbon values, objects, storage, variables,
+and pointers will work. This includes fleshing out the design for:
+
+-   The expression categories used in Carbon to represent values and objects,
+    how they interact, and terminology that anchors on their expression nature.
+-   An expression category model for readonly, abstract values that can
+    efficiently support function inputs.
+-   A customization system for value expression representations, especially as
+    seen on function boundaries in the calling convention.
+-   An expression category model for references instead of a type system model.
+-   How patterns match different expression categories.
+-   How initialization works in conjunction with function returns.
+-   Specific pointer syntax, semantics, and library customization mechanisms.
+-   A `const` type qualifier for use when the value expression category system
+    is too abstracted from the underlying objects in storage.
+
+## Problem
+
+Carbon needs a design for how values, variables, objects, pointers, and
+references work within the language. These designs are heavily interdependent
+and so they are presented together here. The design also needs to provide a
+compelling solution for a wide range of use cases in the language:
+
+-   Easy to use, and use correctly, function input and in/out parameters.
```

Comment:
The items in this list do not have parallel structure. Either they should all look like nouns or they all should have verbs.

Reply:
@jonmeow @wolffg could you help us with this documentation style question?

Reply:
FWIW, this is just in the proposal, so resolving for now. I don't think we need to micro-optimize this.